vault/programs/vault/src/lib.rs

```diff
@@ -0,0 +1,464 @@
+            bump = ctx.accounts.vault.bumps.vault_payer
+        );
+        zeta_client::deposit(
+            ctx.accounts.zeta_program.clone(),
+            ctx.accounts.deposit_cpi_accounts.clone(),
+            seeds,
+            amount,
+        )
+    }
+
+    // #[access_control(deposit_withdraw_phase(&ctx.accounts.vault))]
+    pub fn withdraw_zeta(ctx: Context<WithdrawZeta>, amount: u64) -> ProgramResult {
+        let vault_name = ctx.accounts.vault.vault_name.as_ref();
+        let seeds = vault_payer_seeds!(
+            vault_name = vault_name,
+            bump = ctx.accounts.vault.bumps.vault_payer
+        );
+        zeta_client::withdraw(
+            ctx.accounts.zeta_program.clone(),
+            ctx.accounts.withdraw_cpi_accounts.clone(),
+            seeds,
+            amount,
+        )
+    }
+
+    pub fn initialize_zeta_open_orders(ctx: Context<InitializeZetaOpenOrders>) -> ProgramResult {
+        let vault_name = ctx.accounts.vault.vault_name.as_ref();
+        let seeds = vault_payer_seeds!(
+            vault_name = vault_name,
+            bump = ctx.accounts.vault.bumps.vault_payer
+        );
+        zeta_client::initialize_open_orders(
+            ctx.accounts.zeta_program.clone(),
+            ctx.accounts.initialize_open_orders_cpi_accounts.clone(),
+            seeds,
+        )
+    }
+
+    // TODO: in future move this on-chain
+    pub fn place_auction_order(
+        ctx: Context<PlaceAuctionOrder>,
+        price: u64,
+        size: u64,
+        side: Side,
+        client_order_id: Option<u64>,
+    ) -> ProgramResult {
+        msg!("PLACE AUCTION ORDER");
+        let vault_name = ctx.accounts.vault.vault_name.as_ref();
+        let seeds = vault_payer_seeds!(
+            vault_name = vault_name,
+            bump = ctx.accounts.vault.bumps.vault_payer
+        );
+        zeta_client::place_order(
+            ctx.accounts.zeta_program.clone(),
+            ctx.accounts.place_order_cpi_accounts.clone(),
+            seeds,
+            price,
+            size,
+            side,
+            client_order_id,
+        )
+    }
+
+    pub fn cancel_auction_order(
+        ctx: Context<CancelAuctionOrder>,
+        side: Side,
+        order_id: u128,
+    ) -> ProgramResult {
+        msg!("CANCEL AUCTION ORDER");
+        let vault_name = ctx.accounts.vault.vault_name.as_ref();
+        let seeds = vault_payer_seeds!(
+            vault_name = vault_name,
+            bump = ctx.accounts.vault.bumps.vault_payer
+        );
+        zeta_client::cancel_order(
+            ctx.accounts.zeta_program.clone(),
+            ctx.accounts.cancel_order_cpi_accounts.clone(),
+            seeds,
+            side,
+            order_id,
+        )
+    }
+
+    #[access_control(epoch_over(&ctx.accounts.vault))]
+    pub fn rollover_vault(
+        ctx: Context<RolloverVault>,
+        _vault_name: String,
+        _bumps: VaultBumps,
+        epoch_times: EpochTimes,
```

Comment:
Can these new times be determined trustlessly from the previous epoch times?

Reply:
Most certainly, I was hesitant to put them in because I didn't want to lock people into predetermined 1 week schedules just yet. @epsilonsized thoughts on whether we'd want to enforce it from the program side?

Reply:
So options are:

1. Vault passes in it's own epoch t+1 timings from the client side, little or no validation done other than times being sequential (but theoretically you could change it to a 2 weeker and start it mid week or whatever). So pretty flexible
2. Lock in to a predetermined cadence, calling rollover() crank will just roll it over to the next set of predetermined epoch times - no intervention required at all

Reply:
2) seems cool to do later on because its q trustless... let's stick with 1 now though as it just gives more optionality without having to put up a new vault instance each time.

Reply:
Going with option 2